### ISSUE TITLE
Cannot execute certain shell commands via cli.Client

### ISSUE THREAD

**Issue by Ichimonji10** [PulpQE/pulp-smash#89]
Python's `subprocess` module offers a very handy function under Python 3, `run`. It's used much like `cli.Client.run`. One of the few reasons I'm not currently using it is that it's unavailable on Python 2. Interest in Python 2 is also why I backported(ish) `pulp_smash.cli.CompletedProcess`. Here's a usage example:

``` python
>>> subprocess.run('echo foo'.split())
foo
CompletedProcess(args=['echo', 'foo'], returncode=0)
```

By default, `subprocess.run` does not allow you to use typical shell features like redirection:

``` python
>>> subprocess.run('echo foo > bar'.split())
foo > bar
CompletedProcess(args=['echo', 'foo', '>', 'bar'], returncode=0)
```

However, there's an escape hatch: you can ask for your command to be executed via a system shell. By default, it's `/bin/sh`. Here's an example:

``` python
>>> subprocess.run('echo foo > bar', shell=True)  # creates `bar`!
CompletedProcess(args='echo foo > bar', returncode=0)
```

Now, `cli.Client.run` lets you run commands in a manner similar to `subprocess.run`. It acts just like `subprocess.run` in most cases, even down to preventing redirection by default:

``` python
>>> client.run('echo foo > bar'.split())
CompletedProcess(args=['echo', 'foo', '>', 'bar'], returncode=0, stdout='foo > bar\n', stderr='')
```

However, the escape hatch is broken. We're trapped, and we cannot do things like pipe values. Here's an example where I try to do so, but instead get a `CommandNotFound` exception complaining about the missing command "e" (presumably because `''.split()` is called on the string):

``` python
>>> client.run('echo foo > bar')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/ichimonji10/code/pulp-smash/pulp_smash/cli.py", line 199, in run
    code, stdout, stderr = self.machine[args[0]].run(args[1:], **kwargs)
  File "/home/ichimonji10/.virtualenvs/pulp-smash3/lib/python3.5/site-packages/plumbum/machines/local.py", line 208, in __getitem__
    return LocalCommand(self.which(cmd))
  File "/home/ichimonji10/.virtualenvs/pulp-smash3/lib/python3.5/site-packages/plumbum/machines/local.py", line 178, in which
    raise CommandNotFound(progname, list(cls.env.path))
plumbum.commands.processes.CommandNotFound: ('e', [<LocalPath /home/ichimonji10/.virtualenvs/pulp-smash3/bin>, <LocalPath /usr/local/sbin>, <LocalPath /usr/local/bin>, <LocalPath /usr/bin>, <LocalPath /usr/lib/jvm/default/bin>, <LocalPath /usr/bin/site_perl>, <LocalPath /usr/bin/vendor_perl>, <LocalPath /usr/bin/core_perl>])
```

The core issue is that we're using Plumbum. Plumbum's use case is to provide convenient shell-like combinators by overriding a bunch of Python built-ins, like so:

``` python
from plumbum import local
from plumbum.cmd import wc, ls, echo, grep
from plumbum.utils import copy, delete

delete(local.cwd // "*/*.pyc")
for fn in local.cwd / "src" // "*.py":
    print wc("-l", fn)
num_of_src_lines = (ls["-l"] | grep["\\.py"] | wc["-l"])()
(echo["1"] > "/proc/sys/net/ipv4/ip_forward")()
```

I've not been doing that. I've been using it because it offers an `SshMachine` class, and that's it.

What's the consequence? The consequence is that I cannot write certain kinds of CLI tests. For example, this is an absolutely necessary sequence of commands to test pulp-admin's errata creation functionality:

``` sh
pkglist_csv=$(mktemp)
echo bear,4.1,1,0,noarch,... > "$pkglist_csv"
```

The above cannot be accomplished with `cli.Client`, due to the unavailability of redirection.

Let's drop the middleman, Plumbum, and make `cli.Client` interface with `subprocess.run` ourselves. Well, that's what I'd like to do. But that means dropping support for Python 2.

**Comment by omaciel** [comment]
Tough call...

**Comment by Ichimonji10** [comment]
Perhaps this is a bug in Plumbum that can be fixed.

**Comment by Ichimonji10** [comment]
See toy code at: https://github.com/PulpQE/pulp-smash/compare/master...Ichimonji10:test-errata

**Comment by omaciel** [comment]
Clever 😀

**Comment by Ichimonji10** [comment]
@elyezer You had a possible solution for this. Can you summarize the solution here?

**Comment by elyezer** [comment]
Plumbum is designed so it can allow some smart commands piping and I/O redirection using Python's `|`, `>`, `>>` operators but this require that a command instance be created, for example:

```
grep = client.machine['grep']
echo = client.machine['echo']

(echo["Something"] | grep['Something'])()
```

That is great when you are able to define the commands somewhere and then reuse.

On Pulp Smash tests something a basic piping or I/O redirection is needed and instead of creating the command library with many different arguments is easier to access Plumbum machine session and do the piping or I/O redirection passing as a string like:

```
client.machine.session().run('echo Something | grep Something') which is clean and easy to do.

**Comment by kersommoura** [comment]
Updating the previous link for the sample usage. 

[Piping commands](https://github.com/PulpQE/pulp-smash/blob/121da67b4b12cacfd611f53b7229a3bd71f204b4/pulp_smash/tests/pulp2/rpm/api_v2/test_content_sources.py#L138)

Besides that, Pulp-Smah dropped support to [Python 2](https://github.com/PulpQE/pulp-smash/commit/07db262f8ff099d4444210fe7eef48b1cd5a6626l) a time ago.

Based on that, I am closing this issue, and feel free to re-open in case of need.
